ndmg/utils/cloud_utils.py

```diff
@@ -182,16 +193,30 @@ def s3_push_data(bucket, remote, outDir, modifier, creds=True, debug=True):
     debug : bool, optional
         Whether to not to push intermediate files created by the pipeline, by default True
     """
-    # TODO : use boto3 for this instead
-    cmd = (
-        'aws s3 cp --exclude "tmp/*" {} s3://{}/{}/{}/{}/ --recursive --acl public-read'
-    )
-    dataset = remote.split("/")[0]
-    rest_of_path_list = remote.split("/")[1:]
-    rest_of_path = os.path.join(*rest_of_path_list)
-    cmd = cmd.format(outDir, bucket, dataset, modifier, rest_of_path)
-    if not creds:
-        print("Note: no credentials provided, may fail to push big files.")
-        cmd += " --no-sign-request"
-    print("Pushing results to S3: {}".format(cmd))
-    subprocess.check_output(cmd, shell=True)
+
+    if creds:
+        # get client with credentials if they exist
+        try:
+            client = s3_client(service="s3")
+        except:
+            client = boto3.client("s3")
+    else:
+        client = boto3.client("s3")
+    
+    # check that bucket exists
+    bkts = [bk["Name"] for bk in client.list_buckets()["Buckets"]]
+    if bucket not in bkts:
+        sys.exit(
+            "Error: could not locate bucket. Available buckets: " + ", ".join(bkts)
+        )
+
+    # List all files and upload
+    for r, d, f in os.walk(outDir):
+        for file in f:
+            if not 'tmp/' in r: # exclude things in the tmp/ folder
+                print(f'Uploading: {os.path.join(r, file)}')
+                d = r.replace(os.path.join('/',*outDir.split('/')[:-2]),'') #remove everything before /sub-*
+                client.upload_file(
+                    os.path.join(r, file), bucket, f'{remote}/{modifier}{os.path.join(d,file)}',
+                    ExtraArgs={'ACL':'public-read'}
+                    )
```

Comment:
stylistically, I'm usually a fan of defining variables outside of function calls. Takes up an extra line, but then you can use the variable elsewhere, and I think the name makes the function call a bit easier to understand